README.md
# Native HLS Plugin #

A wrapper for browsers that supports natively HLS that adds additional control and configuration over default native playback.

### How do I get set up? ###

When initializing the player add `NativeHls` with it's own configuration to the configuration object.

``` JavaScript
var meisterPlayer = new Meister("#player", {
    NativeHls: {
        filterAudioOnly: true,
    },
});

meisterPlayer.setItem({
    src: 'INSERT_HLS_M3U8_HERE',
    type: 'm3u8',
});
```

### Configuration ###

Options are required unless marked as [optional].

* [optional] [**filterAudioOnly**](#filteraudioonly-boolean-default-false)
* [optional] [**dvrEnabled**](#dvrenabled-boolean-default-true)
* [optional] [**dvrThreshold**](#dvrthreshold-number-default-300)
* [optional] [**safariDesktopDisabled**](#safaridesktopdisabled-boolean-default-false)

Config options
-------

#### filterAudioOnly *[Boolean]* (default: false) ####

When set to true the player will filter out stream that only contain audio. This means they will not be visible in the ui to select.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    NativeHls: {
        filterAudioOnly: true,
    }
});
```

#### dvrEnabled *[Boolean]* (default: true) ####

Setting this flag to true will treat a HLS stream with DVR window as a livestream.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    NativeHls: {
        dvrEnabled: false,
    }
});
```

#### dvrThreshold *[Number]* (default: 300) ####

This value in seconds is used to determine whether a HLS stream is considered to have a DVR window or not. By default streams that have a window of less than 5 minutes are considered 'just' live.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    NativeHls: {
        dvrThreshold: 50,
    }
});
```

#### safariDesktopDisabled *[Boolean]* (default: false) ####

This flag can be used to disable the plugin in Safari on macOS. This can be useful when you do want to use it for mobile playback, but have a more feature rich HLS plugin for macOS.

Example:

``` JavaScript
var meisterPlayer = new Meister('#player', {
    NativeHls: {
        safariDesktopDisabled: true,
    }
});
```

### startFromBeginning *[Boolean]* (default: false) ###

Start from the beginning of the live stream. (VOD streams will always begin from the beginning).

``` JavaScript
meisterPlayer.setItem({
    src: 'https://example.com/secure/stream/manifest.m3u8',
    startFromBeginning: true,
    type: 'm3u8'
});
```

Or you can give an object with an offset to start from a offset

``` JavaScript
meisterPlayer.setItem({
    src: 'https://example.com/secure/stream/manifest.m3u8',
    startFromBeginning: {
        offset: 10, // Start from the "beginning" with an offset of 10 seconds
    },
    type: 'm3u8'
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [5.6.4](https://github.com/meisterplayer/media-nativehls/compare/v5.6.3...v5.6.4) (2019-06-24)


### Bug Fixes

* **seek:** Fix seeking to the start of a DVR window on load ([25245c0](https://github.com/meisterplayer/media-nativehls/commit/25245c0))



### [5.6.3](https://github.com/meisterplayer/media-nativehls/compare/v5.6.2...v5.6.3) (2019-06-04)


### Bug Fixes

* **live:** Remove arbitrary 30 second buffer when seeking to live ([013abea](https://github.com/meisterplayer/media-nativehls/commit/013abea))



<a name="5.6.2"></a>
## [5.6.2](https://github.com/meisterplayer/media-nativehls/compare/v5.6.1...v5.6.2) (2019-02-06)


### Bug Fixes

* **bitrates:** Fix issue with bitrate availability ([1206879](https://github.com/meisterplayer/media-nativehls/commit/1206879))



<a name="5.6.1"></a>
## [5.6.1](https://github.com/meisterplayer/media-nativehls/compare/v5.6.0...v5.6.1) (2018-02-22)


### Bug Fixes

* **autoplay:** Start playback correctly when an ad ends ([7b77bfa](https://github.com/meisterplayer/media-nativehls/commit/7b77bfa))
* **hls-ads-drm:** fix for Safari Mac and Safari iOS not playing live hls stream after preroll ([c33fa05](https://github.com/meisterplayer/media-nativehls/commit/c33fa05))



<a name="5.6.0"></a>
# [5.6.0](https://github.com/meisterplayer/media-nativehls/compare/v5.5.1...v5.6.0) (2018-02-05)


### Features

* **drm:** Extract DRM information from manifest ([2cc6004](https://github.com/meisterplayer/media-nativehls/commit/2cc6004))



<a name="5.5.1"></a>
## [5.5.1](https://github.com/meisterplayer/media-nativehls/compare/v5.5.0...v5.5.1) (2017-09-05)


### Bug Fixes

* **events:** Fix event names for id3Tags ([31c24e6](https://github.com/meisterplayer/media-nativehls/commit/31c24e6))



<a name="5.5.0"></a>
# [5.5.0](https://github.com/meisterplayer/media-nativehls/compare/v5.4.0...v5.5.0) (2017-09-04)


### Features

* **track:** Add support for reading id3 tags ([3759fbe](https://github.com/meisterplayer/media-nativehls/commit/3759fbe))



<a name="5.4.0"></a>
# [5.4.0](https://github.com/meisterplayer/media-nativehls/compare/v5.3.0...v5.4.0) (2017-08-21)


### Features

* **startFromBeginning:** Add offset support ([9576478](https://github.com/meisterplayer/media-nativehls/commit/9576478))



<a name="5.3.0"></a>
# [5.3.0](https://github.com/meisterplayer/media-nativehls/compare/v5.2.0...v5.3.0) (2017-07-13)


### Features

* **time:** Add duration and currentTime property accessors ([8cbf113](https://github.com/meisterplayer/media-nativehls/commit/8cbf113))



<a name="5.2.0"></a>
# [5.2.0](https://github.com/meisterplayer/media-nativehls/compare/v5.1.0...v5.2.0) (2017-06-27)


### Features

* **pluginVersion:** Add pluginVersion to class ([799ce7e](https://github.com/meisterplayer/media-nativehls/commit/799ce7e))



<a name="5.1.0"></a>
# [5.1.0](https://github.com/meisterplayer/media-nativehls/compare/v5.0.2...v5.1.0) (2017-06-21)


### Features

* **startFromBeginning:** Add option to startFromBeginning ([04d5d51](https://github.com/meisterplayer/media-nativehls/commit/04d5d51))



<a name="5.0.2"></a>
## 5.0.2 (2017-05-09)


### Bug Fixes

* **dvrThreshold:** Fix issue where a 0 threshold was not set ([bfadf5a](https://github.com/meisterplayer/media-nativehls/commit/bfadf5a))
* **userAgent:** Fix issue where Facebook inline browser would not play HLS on iOS ([9a88d0b](https://github.com/meisterplayer/media-nativehls/commit/9a88d0b))


tests/NativeHls.test.js
import '@meisterplayer/meister-mock';
import NativeHls from '../src/js/NativeHls';

const PLUGIN_NAME = 'NativeHls';
const SUPPORTED_TYPES = ['m3u8', 'm3u'];

describe('Hls class', () => {
    test(`pluginName should be ${PLUGIN_NAME}`, () => {
        expect(NativeHls.pluginName).toBe(PLUGIN_NAME);
    });

    test('pluginVersion should return a version string', () => {
        // Version should match the SemVer pattern (e.g. 2.11.9)
        expect(NativeHls.pluginVersion).toMatch(/\d+\.\d+\.\d+/);
    });
});


gulpfile.js
const gulp = require('gulp');
const tripleGulp = require('meister-js-dev').gulp;
const webpackTask = require('meister-gulp-webpack-tasks');

// Building tasks.
const MODULE_NAME = 'NativeHls';

gulp.task('build', (done) => {
    const bundleConfig = webpackTask.createConfig('./index.js', `build/${MODULE_NAME}.js`, false);
    const bundleCompiler = webpackTask.createCompiler(bundleConfig);

    webpackTask.createBuildTask(bundleCompiler)(done);
});

gulp.task('build:dist', (done) => {
    const bundleConfigDist = webpackTask.createConfig('./index.js', `dist/${MODULE_NAME}.js`, false);
    const bundleCompilerDist = webpackTask.createCompiler(bundleConfigDist);

    webpackTask.createBuildTask(bundleCompilerDist)(done);
});

gulp.task('build:min', (done) => {
    const bundleConfigMin = webpackTask.createConfig('./index.js', `dist/${MODULE_NAME}.min.js`, true);
    const bundleCompilerMin = webpackTask.createCompiler(bundleConfigMin);

    webpackTask.createBuildTask(bundleCompilerMin)(done);
});

// Documentation tasks.
gulp.task('js-docs', tripleGulp.jsdocModule.createGenerateDocs(['./src/**/*.js'], './docs/js-docs'));

// Versioning tasks.
gulp.task('bump-version', tripleGulp.versioningModule.createBumpVersion('./package.json'));

// Changelog tasks.
gulp.task('changelog', tripleGulp.changelogModule.createGenerateLog('./CHANGELOG.md'));

// Cleaning tasks
gulp.task('clean:build', tripleGulp.cleanModule.createClean('./build/*'));
gulp.task('clean:dist', tripleGulp.cleanModule.createClean('./dist/*'));


src/js/utils/Http.js
/**
 * @class httpModule
 * @type {{getPageByURL, handleResponse, parseCookie, queryStringToJsonObject, jsonObjectToQuerystring}}
 */

var _totalMillisecondsDelta = 0,
    _noop = function () {},
    _activeXHRs = [],
    _totalBytesDelta = 0,
    _maximumHistorySize = 7,
    _totalBytesDeltaHistory = [],
    _totalMillisecondsDeltaHistory = [],
    _isObjLiteral = function (_obj) {
        var _test = _obj;
        return (typeof _obj !== 'object' || _obj === null ?
            false :
            (
                (function () {
                    while (!false) {
                        if (Object.getPrototypeOf(_test = Object.getPrototypeOf(_test)) === null) {
                            break;
                        }
                    }
                    return Object.getPrototypeOf(_obj) === _test;
                })()
            )
        );
    },
    _verifiedOptions = function (options) {
        if (!_isObjLiteral(options)) {
            console.error('Cannot construct http request, the options parameter is not an object literal.', options);
            return false;
        }

        if (!('method' in options)) {
            console.error('Cannot construct http request, the method was not specified.', options);
            return false;
        }

        if (!('url' in options)) {
            console.error('Cannot construct http request, the url was not specified.', options);
            return false;
        }

        return true;
    },
    _objectToQueryString = function (obj) {
        var parts = [];
        for (var key in obj) {
            if (obj.hasOwnProperty(key)) {
                parts.push(encodeURIComponent(key) + '=' + encodeURIComponent(obj[key]));
            }
        }
        return parts.join('&');
    },
    _updateProgress = function (event) {
        // Make sure to reset the bytes downloaded (can happen after an abort).
        if (event.loaded < this.bytesDownloaded) {
            this.bytesDownloaded = 0;
        }

        // Get the number of new bytes downloaded.
        var bytesDelta = event.loaded - this.bytesDownloaded,
            millisecondsDelta = event.timeStamp - this.timestamp;

        // Update the total values.
        _totalMillisecondsDelta += millisecondsDelta;
        _totalBytesDelta += bytesDelta;
        Http.totalBytesDownloaded += bytesDelta;

        // Store local values.
        this.bytesDownloaded = event.loaded;
        this.totalSizeInBytes = event.total;
        this.timestamp = event.timeStamp;

        // Update the progress.
        this.downloadProgress = this.bytesDownloaded / this.totalSizeInBytes;

        Http.resetStatistics();
    };

/**
 * Makes an HTTP request and serves as an advanced HTTP Api
 *
 * @param  {[type]}   options  [description]
 * @param  {Function} callback [description]
 * @return {[type]}            [description]
 */
function Http(options, callback) {

    if (!(_verifiedOptions(options))) {
        return;
    }

    this.instID = this.constructor.name + '.' + (Math.random() * (new Date().getTime()));

    var method = options.method,
        url = options.url,
        data = '',
        timeout = 'timeout' in options ? options.timeout : 3000,
        xhr;

    this.timestamp = 0;
    this.method = method;
    this.url = url;

    this.status = 0;
    this.bytesDownloaded = 0;
    this.totalSizeInBytes = 0;
    this.downloadProgress = 0;
    this.bandwidth = null;
    this.timeOfRequest = null;
    this.timeOfResponse = null;
    this.roundTripTime = null;
    this.onReady = null;
    this.onProgress = null;
    this.onLoadStart = null;
    this.onError = null;
    this.onSuccess = callback;

    xhr = new XMLHttpRequest();

    if ('data' in options) {
        data = _objectToQueryString(options.data);
    }

    xhr.onreadystatechange = function (event) {
        this.status = xhr.status;

        if (this.onReady !== null) {
            this.onReady();
        }
    }.bind(this);

    xhr.onprogress = function (event) {
        _updateProgress.call(this, event);

        if (this.onProgress !== null) {
            this.onProgress();
        }
    }.bind(this);

    xhr.onloadstart = function (event) {
        this.timestamp = event.timeStamp;

        if (this.onLoadStart !== null) {
            this.onLoadStart();
        }
    }.bind(this);

    xhr.onerror = function (event) {
        Http.downloadingRequests -= 1;

        if (this.onError !== null) {
            (this.onError || _noop)(xhr);
        } else {
            this.onSuccess(null);
        }
    }.bind(this);

    xhr.onload = function (event) {
        _updateProgress.call(this, event);

        Http.downloadingRequests -= 1;

        // Remove the active XHR from active XHRs
        for (var i = 0; i < _activeXHRs.length; i++) {
            var activeXHR = _activeXHRs[i];


            if (activeXHR.id == this.instID) {
                _activeXHRs.splice(i, 1);
            }
        }

        this.timeOfResponse = (new Date()).getTime();
        this.roundTripTime = this.timeOfResponse - this.timeOfRequest;
        xhr.bandwidth = Math.floor((this.bytesDownloaded / this.roundTripTime) * 8 * 1000);
        this.bandwidth = xhr.bandwidth;
        this.onSuccess(xhr);
    }.bind(this);

    // GET requests should have the data in the URL instead of the body
    if (method === 'GET') {
        url += data;
    }

    // Open the asynchronous request
    xhr.open(method, url, true);

    this.timeOfRequest = (new Date()).getTime();
    Http.downloadingRequests += 1;

    //Set the responseType
    if ('responseType' in options) {
        xhr.responseType = options.responseType;
    }

    // Add headers, must be done after open but before send (see mdn docs)
    if ('headers' in options) {
        for (var key in options.headers) {
            if (options.headers.hasOwnProperty(key)) {
                console.log('Setting header', key, 'with', options.headers[key]);
                xhr.setRequestHeader(key, options.headers[key]);
            }
        }
    }

    _activeXHRs.push({
        id: this.instID,
        inst: this
    });

    try {
        xhr.send(data);
    } catch (ex) {
        console.error('HTTP Request aborted due: ', ex);
        (this.onError || _noop)(xhr);
    }


    // Listen to timeouts.
    this.timeout = setTimeout(function () {
        if (this.bytesDownloaded === 0) {
            this.abort();

            if (this.onError !== null) {
                this.onError(xhr);
            } else {
                this.onSuccess(null);
            }
        }

        clearTimeout(this.timeout);
    }.bind(this), timeout);

    this.xhr = xhr;
}

Http.downloadingRequests = 0;
Http.totalBytesDownloaded = 0;
Http.averagedBandwidth = 0;

Http.get = function (url, callback) {
    var options = {
        url: url,
        method: 'GET',
        data: {}
    };

    return new Http(options, callback);
};

Http.post = function (url, data, callback) {
    var options = {
        url: url,
        method: 'POST',
        data: data
    };

    return new Http(options, callback);
};

Http.abortAll = function () {
    for (var i = 0; i < _activeXHRs.length; i++) {
        var activeXHR = _activeXHRs[i];
        activeXHR.inst.abort();
    }
};

Http.resetStatistics = function () {
    var milliseconds = 0.0,
        bytes = 0.0,
        avgBandwidth = null,
        i;

    // Push the total milliseconds delta if there is data.
    if (_totalMillisecondsDelta > 0) {
        _totalMillisecondsDeltaHistory.push(_totalMillisecondsDelta);
        _totalMillisecondsDelta = 0;

        if (_totalMillisecondsDeltaHistory.length > _maximumHistorySize) {
            _totalMillisecondsDeltaHistory.shift();
        }
    }

    // Push the total bytes delta if there is data.
    if (_totalBytesDelta > 0) {
        _totalBytesDeltaHistory.push(_totalBytesDelta);
        _totalBytesDelta = 0;
        if (_totalBytesDeltaHistory.length > _maximumHistorySize) {
            _totalBytesDeltaHistory.shift();
        }
    }

    // Calculate the total of milliseconds.
    for (i = 0; i < _totalMillisecondsDeltaHistory.length; i++) {
        milliseconds += _totalMillisecondsDeltaHistory[i];
    }

    // Calculate the total of bytes.
    for (i = 0; i < _totalBytesDeltaHistory.length; i++) {
        bytes += _totalBytesDeltaHistory[i];
    }

    // Calculate the average bandwidth.
    if (milliseconds > 0) {
        avgBandwidth = (bytes * 1000) / milliseconds;
    }

    Http.averagedBandwidth = Http.averagedBandwidth === 0 ? avgBandwidth : (Http.averagedBandwidth + avgBandwidth) / 2;

    // Reset all values.
    _totalMillisecondsDelta = 0;
    _totalBytesDelta = 0;
    //Http.totalBytesDownloaded = 0;
};

Http.prototype.abort = function () {
    var xhr = this.xhr;
    xhr.onerror = null;

    // Abort the download.
    xhr.abort();

    // Update the number of active requests.
    Http.downloadingRequests -= 1;
};

/**
 * Sugar syntax for when a request is done
 *
 * @param  {Function} callback [description]
 * @return {Function}          [description]
 */
Http.prototype.done = function (callback) {
    this.onSuccess = callback;
    return this;
};

/**
 * Sugar syntax for when a request throws an error
 *
 * @param  {Function} callback [description]
 * @return {[type]}            [description]
 */
Http.prototype.error = function (callback) {
    this.onError = callback;
    return this;
};

/**
 * Sugar syntax for when a request has progress
 *
 * @param  {Function} callback [description]
 * @return {[type]}            [description]
 */
Http.prototype.progress = function (callback) {
    this.onProgress = callback;
    return this;
};

module.exports = Http;


src/js/NativeHls.js
import M3u8Parser from './utils/M3u8Parser';
import packageJson from '../../package.json';
import isAdItem from './utils/isAdItem';
import Id3Tag from './models/Id3Tag';

const POLL_INTERVAL = 1000;

class NativeHls extends Meister.MediaPlugin {
    constructor(config, meister, next) {
        super(config, meister);

        this.manifestParsed = false;

        this.audioMode = false;

        // Middleware promise chain.
        this.next = next;

        // -1 for automatic quality selection
        this.previousLevel = -1;
        this.lowestLevel = 0;

        this.dvrThreshold = this.config.dvrThreshold;

        if (typeof this.dvrThreshold === 'undefined') {
            this.dvrThreshold = 300;
        }

        // new
        this.mediaDuration = 0;
        this.endTime = 0;
        this.beginTime = 0;
        this.mediaSequence = 0;
        this.lastMediaSequence = 0;
        this.childManifest = null;
        this.qualityStreams = [];
        this.masterPlaylist = null;

        this.baseEndTime = 0;

        this.manifestTimeoutId = null;

        // Keep track of the current playback quality.
        this.pollResolutionId = null;
        this.currentResolution = {
            width: 0,
            height: 0,
        };

        this.name = 'nativeHLS';
        this.events = [];
    }

    static get pluginName() {
        return 'NativeHls';
    }

    static get pluginVersion() {
        return packageJson.version;
    }

    isItemSupported(item) {
        return new Promise((resolve) => {
            if (item.type !== 'm3u8' && item.type !== 'm3u') {
                resolve({
                    supported: false,
                    errorCode: Meister.ErrorCodes.WRONG_TYPE,
                });
            }

            // Non Safari browsers are not supported.
            if ((!this.meister.browser.isSafari)
                // When safariDesktopDisabled is true and it's not iOS disable it.
                || (!this.meister.browser.isiOS && this.config.safariDesktopDisabled)
            ) {
                // Make sure we are not in Facebook mode. Where the User agent is modified.
                // This means we are Safari on iOS but no indication of it in the user agent.
                // Allow these browsers to pass.
                if (!this.meister.browser.isFacebook && !this.meister.browser.isiOS) {
                    resolve({
                        supported: false,
                        errorCode: Meister.ErrorCodes.NOT_SUPPORTED,
                    });
                }
            }

            if (item.drm || item.drmConfig) {
                this.meister.one('drmKeySystemSupport', (supportedDRMSystems) => {
                    let supported = false;
                    Object.keys(supportedDRMSystems).forEach((key) => {
                        if ((key === 'com.apple.fps' && supportedDRMSystems[key]) ||
                            (key === 'com.apple.fps.1_0' && supportedDRMSystems[key]) ||
                            (key === 'com.apple.fps.2_0' && supportedDRMSystems[key])) {
                            supported = true;
                        }
                    });
                    resolve({
                        supported,
                        errorCode: supported ? null : Meister.ErrorCodes.NO_DRM,
                    });
                });

                this.meister.trigger('requestDrmKeySystemSupport', {});
            } else {
                resolve({
                    supported: true,
                });
            }
        });
    }

    resetPrivates() {
        this.previousMetadata = null;

        this.manifestParsed = false;

        this.previousLevel = -1;
        this.lowestLevel = 0;
        this.mediaDuration = 0;
        this.item = null;
    }

    async process(item) {
        try {
            const newItem = await this.next(item);

            this.player = this.meister.getPlayerByType('html5', newItem);
            if (this.meister.config.audioOnly || newItem.mediaType === 'audio') {
                this.audioMode = true;
            } else {
                this.audioMode = false;
            }

            return newItem;
        } catch (err) {
            console.error(`Something went wrong while processing middlewares. ${err}`);
            return null;
        }
    }

    async load(item) {
        super.load(item);
        this.item = item;

        const manifest = await this.loadManifest(item.src);

        this.mediaElement = this.player.mediaElement;

        // // The current playlist item
        const currentPlaylistItem = this.meister.playlist.list[this.meister.playlist.index];
        // PLEASE NOTE:
        // This is NOT the same as is done in DASH or HLS.
        // If the event is handled on Safari MAC (with livestreams and ADS) it won't start playback after the preroll
        // However; safari on iOS won't start if the event isn't handled, so.. nice going Apple :(
        if (isAdItem(currentPlaylistItem) && (this.meister.browser.isMobile && this.meister.browser.isSafari)) {
            this.one('GoogleIma:initialUserActionCompleted', () => {
                this.mediaElement.src = item.src;
            });
        } else {
            this.mediaElement.src = item.src;
        }

        this.masterPlaylist = item.src;

        // Listen to control events.
        this.on('requestBitrate', this.onRequestBitrate.bind(this));
        this.on('requestGoLive', () => this.onRequestGoLive());

        this.mediaElement.textTracks.addEventListener('addtrack', this.onAddTextTrack.bind(this));

        this.pollResolutionId = setInterval(this.pollResolution.bind(this), POLL_INTERVAL);

        // Trigger this to make it look pretty.
        // Loading the first playlist.

        if (manifest.keyInfo) {
            if (manifest.keyInfo.URI) {
                const drmServerUrl = manifest.keyInfo.URI.replace('skd:', 'https:');

                this.meister.trigger('drmLicenseInfoAvailable', {
                    fairplay: {
                        drmServerUrl,
                    },
                });
            }
        }

        this.endTime = manifest.duration;
        this.baseEndTime = this.endTime;
        this.mediaDuration = manifest.duration;
        this.mediaSequence = manifest.mediaSequence;

        this.beginTime = this.endTime - this.mediaDuration;

        // Kinda weird, but let's roll with it for now..
        const lastMediaSequence = Object.keys(manifest.segments)[(Object.keys(manifest.segments).length - 1)];
        this.lastMediaSequence = lastMediaSequence;

        let hasDVR = ((manifest.duration > this.dvrThreshold) && manifest.isLive);

        if (this.config.dvrEnabled === false) {
            hasDVR = false;
        }

        this.meister.trigger('itemTimeInfo', {
            isLive: manifest.isLive,
            hasDVR,
            duration: this.mediaDuration,
            modifiedDuration: this.mediaDuration,
            endTime: this.endTime,
        });

        // We don't want to request live when we want to start from the beginning.
        if (!item.startFromBeginning) {
            // this.onMasterPlaylistLoaded(manifest);
            if (manifest.isLive) this.onRequestGoLive();
        } else if (typeof item.startFromBeginning === 'object') {
            this.onRequestSeek({
                relativePosition: item.startFromBeginning.offset / this.mediaDuration,
            });
        } else if (isNaN(this.meister.duration)) {
            this.meister.one('playerCanPlay', () => {
                this.meister.currentTime = 0;
            });
        } else {
            this.meister.currentTime = 0;
        }

        this.manifestTimeoutId = setTimeout(() => {
            this.getNewManifest();
        }, 5000); // Amount of seconds should be dynamic (By using the manifest)
    }

    get duration() {
        return this.mediaDuration;
    }

    get currentTime() {
        if (!this.player) { return NaN; }

        const playOffset = this.endTime - this.mediaDuration;
        return this.player.currentTime - playOffset;
    }

    set currentTime(time) {
        if (!this.player) { return; }

        const playOffset = this.endTime - this.mediaDuration;
        this.player.currentTime = time + playOffset;
    }

    _onPlayerTimeUpdate() {
        const playOffset = this.endTime - this.mediaDuration;

        this.meister.trigger('playerTimeUpdate', {
            currentTime: this.player.currentTime - playOffset,
            duration: this.mediaDuration,
        });
    }

    _onPlayerSeek() {
        const playOffset = this.endTime - this.mediaDuration;

        const currentTime = this.player.currentTime - playOffset;
        const duration = this.mediaDuration;
        const relativePosition = currentTime / duration;

        this.meister.trigger('playerSeek', {
            relativePosition,
            currentTime,
            duration,
        });
    }

    onRequestSeek(e) {
        let targetTime;

        if (Number.isFinite(e.relativePosition)) {
            const playOffset = this.endTime - this.mediaDuration;
            targetTime = (this.mediaDuration * e.relativePosition) + playOffset;
        } else if (Number.isFinite(e.timeOffset)) {
            targetTime = this.player.currentTime + e.timeOffset;
        } else if (Number.isFinite(e.targetTime)) {
            const playOffset = this.endTime - this.mediaDuration;
            targetTime = e.targetTime + playOffset;
        }

        // Check whether we are allowed to seek forward.
        if (!e.forcedStart && this.blockSeekForward && targetTime > this.player.currentTime) { return; }

        if (Number.isFinite(targetTime)) {
            if (isNaN(this.player.duration)) {
                this.one('playerCanPlay', () => {
                    this.player.currentTime = targetTime;
                });
            } else {
                this.player.currentTime = targetTime;
            }
        }
    }

    /**
     * Event handler for addtrack
     *
     * @param {TrackEvent} event
     * @memberof NativeHls
     */
    onAddTextTrack(event) {
        const track = event.track;
        track.mode = 'hidden';

        track.addEventListener('cuechange', (cueChangeEvent) => {
            /** @type {TextTrack} */
            const target = cueChangeEvent.target;
            const activeCues = Array.from(target.activeCues).filter(cue => cue.type === 'org.id3');
            const id3Tags = activeCues.map(cue => new Id3Tag(cue.value.key, cue.value.data, cue.startTime, cue.endTime));

            this.meister.trigger('id3Tags', id3Tags);
        });
    }

    onRequestGoLive() {
        if (isNaN(this.player.duration)) {
            this.meister.one('playerLoadedMetadata', () => {
                this.onRequestGoLive();
            });
        } else {
            this.player.currentTime = this.endTime - 1;
        }
    }

    onRequestBitrate(e) {
        const previousCurrentTime = this.player.currentTime;
        const wasPlaying = this.meister.playing;

        // Since we're setting a new source we could need updated drm settings.
        if (e.drmConfig) {
            this.meister.trigger('drmConfig', e.drmConfig);
        }

        this.meister.playerPlugin.mediaElement.src = '';

        if (e.bitrateIndex === -1) {
            this.meister.playerPlugin.mediaElement.src = this.masterPlaylist;
        } else {
            this.meister.playerPlugin.mediaElement.src = this.meister.utils.resolveUrl(this.masterPlaylist, this.qualityStreams[e.bitrateIndex].url);
        }

        // Reset playoffset and endtime, since we cleared the source.
        this.playOffset = 0;
        this.endTime = this.baseEndTime;

        this.meister.one('playerLoadedMetadata', () => {
            this.player.currentTime = previousCurrentTime;

            if (wasPlaying) {
                this.meister.play();
            } else {
                this.meister.pause();
            }
        });
    }

    onError(e, data) {
        console.warn(`Error in ${this.name}, type: ${data.details}, will attempt to recover.`);
        if (data.fatal) {
            console.error(`Can not recover from ${data.type}: ${data.details}.`);
        }
    }

    pollResolution() {
        const height = this.mediaElement.videoHeight;
        const width = this.mediaElement.videoWidth;

        if (this.currentResolution.width === width && this.currentResolution.height === height) {
            return;
        }

        const newBitrate = this.qualityStreams.find(stream => stream.resolution && stream.resolution.width === width && stream.resolution.height === height);

        // This can happen while switching streams, no need to notify the player.
        if (!newBitrate) {
            return;
        }

        const newBitrateIndex = this.qualityStreams.indexOf(newBitrate);

        this.meister.trigger('playerAutoSwitchBitrate', {
            newBitrate: parseInt(newBitrate.bandwith, 10),
            newBitrateIndex,
        });

        this.currentResolution = newBitrate.resolution;
    }

    // copypaste from native-hls
    async getNewManifest() {
        try {
            const manifest = await this.loadManifest(this.childManifest);
            const lastMediaSequence = Object.keys(manifest.segments)[(Object.keys(manifest.segments).length - 1)];
            const amountOfNewSegments = lastMediaSequence - this.lastMediaSequence;

            for (let i = 0; i < amountOfNewSegments; i += 1) {
                this.endTime += manifest.segments[Object.keys(manifest.segments)[i]];
            }

            // Just for testing purposes:
            this.mediaDuration = manifest.duration;
            this.beginTime = this.endTime - manifest.duration;
            this.lastMediaSequence = lastMediaSequence;

            let hasDVR = ((manifest.duration > this.dvrThreshold) && manifest.isLive);

            if (this.config.dvrEnabled === false) {
                hasDVR = false;
            }

            this.meister.trigger('itemTimeInfo', {
                isLive: manifest.isLive,
                hasDVR,
                duration: this.mediaDuration,
                modifiedDuration: this.mediaDuration,
                endTime: this.endTime,
            });

            this.manifestTimeoutId = setTimeout(() => {
                this.getNewManifest();
            }, 5000);
        } catch (err) {
            console.warn('WARNING: Could not load manifest, retrying loading manifest.', err);
            this.manifestTimeoutId = setTimeout(() => {
                this.getNewManifest();
            }, 5000);
        }
    }

    // copypaste from native-hls
    async loadManifest(src) {
        const response = await fetch(src);
        const text = await response.text();

        const m3u8 = new M3u8Parser(text);
        const manifest = m3u8.parse();

        if (manifest.streams.length) {
            if (this.config.filterAudioOnly) {
                this.qualityStreams = manifest.streams.filter(stream => stream.resolution);
            } else {
                this.qualityStreams = manifest.streams;
            }

            this.onQualitysAvailable();

            this.childManifest = this.meister.utils.resolveUrl(src, manifest.streams[0].url);
            // This is the master playlist we need to parse the sub playlist.
            return this.loadManifest(this.meister.utils.resolveUrl(src, manifest.streams[0].url));
        }

        return manifest;
    }

    // copypaste from native-hls
    onQualitysAvailable() {
        const bitrates = this.qualityStreams.map((bitrate, index) => ({
            bitrate: parseInt(bitrate.bandwith, 10),
            index,
        }));

        // Bitrate 0 means auto quality.
        bitrates.unshift({
            bitrate: 0,
            index: -1,
        });

        // Trigger auto bitrate by default.
        this.meister.trigger('itemBitrates', {
            bitrates,
            currentIndex: -1,
        });

        // Since Conviva requires a bitrate before playing (and we do not know the bitrate)
        // we temporarily set the first bitrate in the manifest. This follows most HLS implementations.
        const firstBitrate = bitrates.find(bitrate => bitrate.index === 0);

        if (!firstBitrate) {
            return;
        }

        this.meister.trigger('playerAutoSwitchBitrate', {
            newBitrate: firstBitrate.bitrate,
            newBitrateIndex: firstBitrate.index,
        });
    }

    unload() {
        super.unload();

        if (this.manifestTimeoutId) {
            clearTimeout(this.manifestTimeoutId);
        }

        if (this.pollResolutionId) {
            clearInterval(this.pollResolutionId);
        }

        this.meister.remove(this.events);

        this.mediaDuration = 0;
        this.endTime = 0;
        this.baseEndTime = 0;
        this.beginTime = 0;
        this.mediaSequence = 0;
        this.lastMediaSequence = 0;
        this.childManifest = null;
        this.qualityStreams = [];
        this.masterPlaylist = null;
        this.manifestTimeoutId = null;
        this.pollResolutionId = null;
        this.currentResolution = {
            width: 0,
            height: 0,
        };
        this.mediaElement = null;
    }

    destroy() {
        super.destroy();
    }
}

Meister.registerPlugin(NativeHls.pluginName, NativeHls);
Meister.registerPlugin('nativehls', NativeHls);

export default NativeHls;
